src/Connection/Connection.ts

```diff
@@ -464,13 +521,76 @@ export class Connection extends Parser<{
      * is invoked.
      */
     private onSocketDisconnect = (): void => {
-        if (!this.teardown) this.emit("Disconnect");
+        if (!this.teardown) {
+            // If not already reconnecting, start the reconnection process
+            if (!this.reconnecting) {
+                this.reconnecting = true;
+                this.attemptReconnect();
+            }
+            this.emit("Disconnect");
+        }
+    };
+    
+    /**
+     * Attempts to reconnect with exponential backoff
+     */
+    private attemptReconnect(): void {
+        if (this.teardown || this.reconnectAttempts >= this.maxReconnectAttempts) {
+            if (this.reconnectAttempts >= this.maxReconnectAttempts) {
+                log.error(`Failed to reconnect after ${this.maxReconnectAttempts} attempts to ${this.host}`);
+            }
+            this.stopReconnectionAttempts();
+            return;
+        }
+
+        const delay = Math.min(
+            30000, // Cap at 30 seconds
+            this.initialBackoffDelay * Math.pow(1.5, this.reconnectAttempts)
+        );
+        
+        log.info(`Attempting reconnection to ${this.host} in ${Math.round(delay / 1000)}s (attempt ${this.reconnectAttempts + 1}/${this.maxReconnectAttempts})`);
+        
+        this.reconnectTimer = setTimeout(() => {
+            this.reconnectAttempts++;
+            
+            // Clear any previous socket
+            this.socket = undefined;
+            
+            log.info(`Reconnecting to ${this.host}...`);
+            this.connect()
+                .then(() => {
+                    log.info(`Successfully reconnected to ${this.host}`);
+                    this.stopReconnectionAttempts();
+                })
+                .catch((error) => {
+                    log.warn(`Failed to reconnect to ${this.host}: ${error.message}`);
+                    // Failed to reconnect, try again
+                    this.attemptReconnect();
+                });
+        }, delay);
     };
 
     /*
      * Listener for any error from the socket.
      */
     private onSocketError = (error: Error): void => {
+        // For network-related errors, attempt reconnection if not already reconnecting
+        if (
+            !this.reconnecting && 
+            !this.teardown && 
+            (
+                error.message.includes('ECONNRESET') || 
```

Comment:
This is a bit difficult to read. This should be moved to a isTimeoutError function.